### ISSUE TITLE
support of pprof files in Perfetto

### ISSUE THREAD

**Issue by nians** [google/perfetto#9]
We are trying to use perfetto as a platform that can analyze and present most of our perf traces, for systrace,  proc stats , they are easy to query and present, but for pprof files, we haven't find a way to handle it

Could you consider implementing the support of pprof files? Thanks very much

**Comment by segfaulthunter** [comment]
Thanks for getting in touch!

Could you elaborate a bit on that? Where do these pprof files come from - from Perfetto or from somewhere else? What would you like us to implement?

**Comment by primiano** [comment]
I think what they mean is being able to import pprof files (not sure if the "perf.data" binary from linux's perf or the gperftools proto) and ingest them in trace processor / ui.
@nians do you have an example file you'd be able to import, or at least the cmdline that generates it?

**Comment by nians** [comment]
@primiano yes, and perf.data is from Android simpleperf, which is same to Linux perf, and there is also tool in simpleperf project can convert perf.data to pprof files, an exampled file is here :[https://drive.google.com/open?id=1JybfYPYUe6OWpJpJQFasMStL5j8gMqP2] and can be shown in speedscope.com

**Comment by caitp** [comment]
It would be handy to be able to do this for chromium tracing as well. it seems perfetto can't deal with the output of pprof-style traces from `--enable-profiling`

**Comment by primiano** [comment]
> It would be handy to be able to do this for chromium tracing as well. it seems perfetto can't deal with the output of pprof-style traces from --enable-profiling


Not sure what you mean with this (you mean the chrome://tracing ui?).
Eventually chrome://tracing will be replaced by ui.perfetto.dev, it's mainly a matter of time. Won't happen too soon though because tehre are various features missing that are blocking.

**Comment by caitp** [comment]
Referring specifically to https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/profiling.md#Profiling-using-built_in-tcmalloc-profiler --- which in my experience, has worked better than perfetto traces. But being able to automate queries over a set of 100 startup traces with the trace processor shell, would be very nice to do. This isn't a thing at the moment.

**Comment by primiano** [comment]
I have no idea how that works but it seems to emit some format that pprof understands, so it's either the google pprof proto or the perf.data.

> which in my experience, has worked better than perfetto traces.

traces and callstack profiles they are two very different things and solve very different classes of problems. This statement makes little sense to me (even if you flip the direction of what works better). It's like saying "this car works way better than this boat"... i mean depends on where you want to go ;)

**Comment by caitp** [comment]
"works better" in the sense that it produces more helpful data for my use case.

**Comment by DanielRosenwasser** [comment]
This would be nice for Node programs that tend to have profile captures bigger than the average web page.

For example, the TypeScript team implemented support for [profiling the compiler](https://github.com/microsoft/TypeScript/wiki/Performance#profiling-the-compiler) to help diagnose our own perf issues. This uses Node's built-in profiling support, but the output of these is *massive*. Instead, we've been using [pprof-nodejs](https://github.com/google/pprof-nodejs) to profile our compiler, but helping users self-service when they're unable to share code involves a lot of steps that we'd like to remove.

Ideally, they could produce output with pprof, and visualize their profiles in Perfetto.

**Comment by primiano** [comment]
@DanielRosenwasser out of curiosity are those pprof profiles symzolized or require symbolization?

It's one of the things @ddiproietto and myself were trying to figure out recently.

**Comment by nolanmar511** [comment]
pprof-nodejs produces fully symbolized profiles (changing JS code should cause issues -- feel free to file an issue with more details). Golang's profiles (generated with net/http/pprof or runtime/pprof) should also be fully symbolized. The `pprof` CLI (https://github.com/google/pprof) does have some utilities for handling unsymbolized profiles; I think this code is pretty much C/C++ profiles (and this symbolization logic requires that the binaries with sufficient debug information are available on the machine which the pprof CLI is running on and requires that those binaries are compiled with sufficient debug information).  

+ @google/pprof-team and + @amchiclet (who is also on the Cloud Profiler team and is the person on our team looking into https://github.com/open-telemetry/oteps/issues/139, which is a proposal for adding profiling into open telemetry).

**Comment by LalitMaganti** [comment]
Support for the perf.data format has been implemented for stack sampling at least. We'll be adding support for counters etc in upcoming releases but I think we've done enough to close this bug.

**Comment by yanj403** [comment]
> Support for the perf.data format has been implemented for stack sampling at least. We'll be adding support for counters etc in upcoming releases but I think we've done enough to close this bug.

Hi Lalit,
You mentioned couters etc will release in upcoming releases, is that means currently perfetto is not support to trace counters like HW cycles for a specific process like we always did in simpleperf with command "simpleperf stat -p <process name>"

**Comment by aMayzner** [comment]
Hi,
Yes, it means that currently perfetto doesn't support any counters, also
those from simpleperf. Feel free to contribute if needed, adding support
for counters should not be too difficult.
